File: dotnet/src/EventGraph/Primitive.cs
namespace EventGraph;

// ── Lifecycle ────────────────────────────────────────────────────────────

public static class Lifecycle
{
    public const string Dormant = "dormant";
    public const string Activating = "activating";
    public const string Active = "active";
    public const string Processing = "processing";
    public const string Emitting = "emitting";
    public const string Suspending = "suspending";
    public const string Suspended = "suspended";
    public const string Memorial = "memorial";

    private static readonly Dictionary<string, HashSet<string>> Transitions = new()
    {
        [Dormant] = new() { Activating },
        [Activating] = new() { Active },
        [Active] = new() { Processing, Suspending, Memorial },
        [Processing] = new() { Emitting, Active },
        [Emitting] = new() { Active },
        [Suspending] = new() { Suspended },
        [Suspended] = new() { Activating, Memorial },
        [Memorial] = new(),
    };

    public static bool IsValidTransition(string from, string to) =>
        Transitions.TryGetValue(from, out var targets) && targets.Contains(to);
}

// ── Mutation types ───────────────────────────────────────────────────────

public abstract record Mutation;
public sealed record AddEventMutation(EventType Type, ActorId Source, Dictionary<string, object?> Content, List<EventId> Causes, ConversationId ConversationId) : Mutation;
public sealed record UpdateStateMutation(PrimitiveId PrimitiveId, string Key, object? Value) : Mutation;
public sealed record UpdateActivationMutation(PrimitiveId PrimitiveId, Activation Level) : Mutation;
public sealed record UpdateLifecycleMutation(PrimitiveId PrimitiveId, string State) : Mutation;

// ── Snapshot ─────────────────────────────────────────────────────────────

public sealed record PrimitiveState(PrimitiveId Id, Layer Layer, string LifecycleState, Activation Activation, Cadence Cadence, Dictionary<string, object?> State, int LastTick);
public sealed record Snapshot(int Tick, Dictionary<string, PrimitiveState> Primitives, List<Event> PendingEvents, List<Event> RecentEvents);

// ── Primitive protocol ───────────────────────────────────────────────────

public interface IPrimitive
{
    PrimitiveId Id { get; }
    Layer Layer { get; }
    List<Mutation> Process(int tick, List<Event> events, Snapshot snapshot);
    List<SubscriptionPattern> Subscriptions { get; }
    Cadence Cadence { get; }
}

// ── Registry ─────────────────────────────────────────────────────────────

public sealed class PrimitiveRegistry
{
    private readonly Lock _lock = new();
    private readonly Dictionary<string, IPrimitive> _primitives = new();
    private readonly Dictionary<string, MutableState> _states = new();
    private List<string> _ordered = new();

    public void Register(IPrimitive p)
    {
        lock (_lock)
        {
            var key = p.Id.Value;
            if (_primitives.ContainsKey(key))
                throw new InvalidOperationException($"Primitive '{key}' already registered");
            _primitives[key] = p;
            _states[key] = new MutableState { Activation = new Activation(0.0), LifecycleState = Lifecycle.Dormant, State = new(), LastTick = 0 };
            RebuildOrder();
        }
    }

    public IPrimitive? Get(PrimitiveId id)
    {
        lock (_lock) { return _primitives.GetValueOrDefault(id.Value); }
    }

    public List<IPrimitive> All()
    {
        lock (_lock) { return _ordered.Select(k => _primitives[k]).ToList(); }
    }

    public int Count { get { lock (_lock) { return _primitives.Count; } } }

    public Dictionary<string, PrimitiveState> AllStates()
    {
        lock (_lock)
        {
            var result = new Dictionary<string, PrimitiveState>();
            foreach (var (key, p) in _primitives)
            {
                var ms = _states[key];
                result[key] = new PrimitiveState(p.Id, p.Layer, ms.LifecycleState, ms.Activation, p.Cadence, new Dictionary<string, object?>(ms.State), ms.LastTick);
            }
            return result;
        }
    }

    public string GetLifecycle(PrimitiveId id)
    {
        lock (_lock) { return _states.TryGetValue(id.Value, out var ms) ? ms.LifecycleState : Lifecycle.Dormant; }
    }

    public void SetLifecycle(PrimitiveId id, string state)
    {
        lock (_lock)
        {
            if (!_states.TryGetValue(id.Value, out var ms))
                throw new InvalidOperationException($"Primitive '{id.Value}' not found");
            if (!Lifecycle.IsValidTransition(ms.LifecycleState, state))
                throw new InvalidTransitionException(ms.LifecycleState, state);
            ms.LifecycleState = state;
        }
    }

    public void Activate(PrimitiveId id)
    {
        lock (_lock)
        {
            if (!_states.TryGetValue(id.Value, out var ms))
                throw new InvalidOperationException($"Primitive '{id.Value}' not found");
            if (!Lifecycle.IsValidTransition(ms.LifecycleState, Lifecycle.Activating))
                throw new InvalidTransitionException(ms.LifecycleState, Lifecycle.Activating);
            ms.LifecycleState = Lifecycle.Active; // dormant -> activating -> active
        }
    }

    public void SetActivation(PrimitiveId id, Activation level)
    {
        lock (_lock)
        {
            if (!_states.TryGetValue(id.Value, out var ms))
                throw new InvalidOperationException($"Primitive '{id.Value}' not found");
            ms.Activation = level;
        }
    }

    public void UpdateState(PrimitiveId id, string key, object? value)
    {
        lock (_lock)
        {
            if (!_states.TryGetValue(id.Value, out var ms))
                throw new InvalidOperationException($"Primitive '{id.Value}' not found");
            ms.State[key] = value;
        }
    }

    public int GetLastTick(PrimitiveId id)
    {
        lock (_lock) { return _states.TryGetValue(id.Value, out var ms) ? ms.LastTick : 0; }
    }

    public void SetLastTick(PrimitiveId id, int tick)
    {
        lock (_lock) { if (_states.TryGetValue(id.Value, out var ms)) ms.LastTick = tick; }
    }

    private void RebuildOrder()
    {
        _ordered = _primitives.Keys
            .OrderBy(k => _primitives[k].Layer.Value)
            .ThenBy(k => k)
            .ToList();
    }

    private sealed class MutableState
    {
        public Activation Activation;
        public string LifecycleState = Lifecycle.Dormant;
        public Dictionary<string, object?> State = new();
        public int LastTick;
    }
}

File: python/eventgraph/primitive.py
"""Primitive protocol, Mutation types, Snapshot, and Registry."""

from __future__ import annotations

import copy
import threading
from dataclasses import dataclass, field
from typing import Any, Protocol, Sequence

from .event import Event
from .types import (
    Activation,
    Cadence,
    EventID,
    EventType,
    Layer,
    PrimitiveID,
    SubscriptionPattern,
)


# ── Lifecycle ────────────────────────────────────────────────────────────

LIFECYCLE_DORMANT = "dormant"
LIFECYCLE_ACTIVATING = "activating"
LIFECYCLE_ACTIVE = "active"
LIFECYCLE_PROCESSING = "processing"
LIFECYCLE_EMITTING = "emitting"
LIFECYCLE_SUSPENDING = "suspending"
LIFECYCLE_SUSPENDED = "suspended"
LIFECYCLE_MEMORIAL = "memorial"

_TRANSITIONS: dict[str, set[str]] = {
    LIFECYCLE_DORMANT: {LIFECYCLE_ACTIVATING},
    LIFECYCLE_ACTIVATING: {LIFECYCLE_ACTIVE},
    LIFECYCLE_ACTIVE: {LIFECYCLE_PROCESSING, LIFECYCLE_SUSPENDING, LIFECYCLE_MEMORIAL},
    LIFECYCLE_PROCESSING: {LIFECYCLE_EMITTING, LIFECYCLE_ACTIVE},
    LIFECYCLE_EMITTING: {LIFECYCLE_ACTIVE},
    LIFECYCLE_SUSPENDING: {LIFECYCLE_SUSPENDED},
    LIFECYCLE_SUSPENDED: {LIFECYCLE_ACTIVATING, LIFECYCLE_MEMORIAL},
    LIFECYCLE_MEMORIAL: set(),
}


def valid_transition(from_state: str, to_state: str) -> bool:
    return to_state in _TRANSITIONS.get(from_state, set())


# ── Mutation types ───────────────────────────────────────────────────────

@dataclass(frozen=True)
class AddEvent:
    type: EventType
    source: Any  # ActorID
    content: dict[str, Any]
    causes: list[EventID]
    conversation_id: Any  # ConversationID


@dataclass(frozen=True)
class UpdateState:
    primitive_id: PrimitiveID
    key: str
    value: Any


@dataclass(frozen=True)
class UpdateActivation:
    primitive_id: PrimitiveID
    level: Activation


@dataclass(frozen=True)
class UpdateLifecycle:
    primitive_id: PrimitiveID
    state: str


Mutation = AddEvent | UpdateState | UpdateActivation | UpdateLifecycle


# ── Snapshot ─────────────────────────────────────────────────────────────

@dataclass(frozen=True)
class PrimitiveState:
    id: PrimitiveID
    layer: Layer
    lifecycle: str
    activation: Activation
    cadence: Cadence
    state: dict[str, Any]
    last_tick: int


@dataclass(frozen=True)
class Snapshot:
    tick: int
    primitives: dict[str, PrimitiveState]  # keyed by primitive_id.value
    pending_events: list[Event]
    recent_events: list[Event]


# ── Primitive protocol ───────────────────────────────────────────────────

class Primitive(Protocol):
    def id(self) -> PrimitiveID: ...
    def layer(self) -> Layer: ...
    def process(self, tick: int, events: list[Event], snapshot: Snapshot) -> list[Mutation]: ...
    def subscriptions(self) -> list[SubscriptionPattern]: ...
    def cadence(self) -> Cadence: ...


# ── Registry ─────────────────────────────────────────────────────────────

class Registry:
    """Thread-safe registry of primitives and their mutable state."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._primitives: dict[str, Primitive] = {}  # keyed by id.value
        self._states: dict[str, _MutableState] = {}
        self._ordered: list[str] = []

    def register(self, p: Primitive) -> None:
        with self._lock:
            key = p.id().value
            if key in self._primitives:
                raise ValueError(f"primitive {key!r} already registered")
            self._primitives[key] = p
            self._states[key] = _MutableState(
                activation=Activation(0.0),
                lifecycle=LIFECYCLE_DORMANT,
                state={},
                last_tick=0,
            )
            self._rebuild_order()

    def get(self, pid: PrimitiveID) -> Primitive | None:
        with self._lock:
            return self._primitives.get(pid.value)

    def all(self) -> list[Primitive]:
        with self._lock:
            return [self._primitives[k] for k in self._ordered]

    def count(self) -> int:
        with self._lock:
            return len(self._primitives)

    def all_states(self) -> dict[str, PrimitiveState]:
        with self._lock:
            result = {}
            for key, p in self._primitives.items():
                ms = self._states[key]
                result[key] = PrimitiveState(
                    id=p.id(),
                    layer=p.layer(),
                    lifecycle=ms.lifecycle,
                    activation=ms.activation,
                    cadence=p.cadence(),
                    state=copy.deepcopy(ms.state),
                    last_tick=ms.last_tick,
                )
            return result

    def lifecycle(self, pid: PrimitiveID) -> str:
        with self._lock:
            ms = self._states.get(pid.value)
            return ms.lifecycle if ms else LIFECYCLE_DORMANT

    def set_lifecycle(self, pid: PrimitiveID, state: str) -> None:
        with self._lock:
            ms = self._states.get(pid.value)
            if ms is None:
                raise ValueError(f"primitive {pid.value!r} not found")
            if not valid_transition(ms.lifecycle, state):
                raise ValueError(
                    f"invalid transition: {ms.lifecycle} -> {state}"
                )
            ms.lifecycle = state

    def activate(self, pid: PrimitiveID) -> None:
        with self._lock:
            ms = self._states.get(pid.value)
            if ms is None:
                raise ValueError(f"primitive {pid.value!r} not found")
            if not valid_transition(ms.lifecycle, LIFECYCLE_ACTIVATING):
                raise ValueError(
                    f"invalid transition: {ms.lifecycle} -> {LIFECYCLE_ACTIVATING}"
                )
            ms.lifecycle = LIFECYCLE_ACTIVATING
            if not valid_transition(LIFECYCLE_ACTIVATING, LIFECYCLE_ACTIVE):
                raise ValueError(
                    f"invalid transition: {LIFECYCLE_ACTIVATING} -> {LIFECYCLE_ACTIVE}"
                )
            ms.lifecycle = LIFECYCLE_ACTIVE

    def set_activation(self, pid: PrimitiveID, level: Activation) -> None:
        with self._lock:
            ms = self._states.get(pid.value)
            if ms is None:
                raise ValueError(f"primitive {pid.value!r} not found")
            ms.activation = level

    def update_state(self, pid: PrimitiveID, key: str, value: Any) -> None:
        with self._lock:
            ms = self._states.get(pid.value)
            if ms is None:
                raise ValueError(f"primitive {pid.value!r} not found")
            ms.state[key] = copy.deepcopy(value)

    def set_last_tick(self, pid: PrimitiveID, tick: int) -> None:
        with self._lock:
            ms = self._states.get(pid.value)
            if ms is not None:
                ms.last_tick = tick

    def last_tick(self, pid: PrimitiveID) -> int:
        with self._lock:
            ms = self._states.get(pid.value)
            return ms.last_tick if ms else 0

    def _rebuild_order(self) -> None:
        self._ordered = sorted(
            self._primitives.keys(),
            key=lambda k: (self._primitives[k].layer().value, k),
        )


class _MutableState:
    __slots__ = ("activation", "lifecycle", "state", "last_tick")

    def __init__(
        self,
        activation: Activation,
        lifecycle: str,
        state: dict[str, Any],
        last_tick: int,
    ) -> None:
        self.activation = activation
        self.lifecycle = lifecycle
        self.state = state
        self.last_tick = last_tick
